src/pages/questions/index.tsx

```diff
@@ -1,42 +1,58 @@
 import Head from "next/head";
-import { useState } from "react";
+import { useEffect, useState } from "react";
 
+import { useSession } from "next-auth/react";
+import { toast } from "react-hot-toast";
+import PageSelector from "~/components/PageSelector";
 import { api } from "~/utils/api";
-import { Question, type QuestionMap } from "../../types/global";
-import { equals } from "../../utils/utils";
 import { QuestionRow } from "../../components/QuestionRow";
 import { StyledButton } from "../../components/StyledButton";
 import { StyledCheckbox } from "../../components/StyledCheckbox";
-import { makeMap } from "../../utils/utils";
-import { toast } from "react-hot-toast";
-import { useSession } from "next-auth/react";
-import WarningModal from "~/components/WarningModal";
-import { WithAuthWrapper } from "~/components/wrapper/AuthWrapper";
+import { Question, type QuestionMap } from "../../types/global";
+import { equals, makeMap } from "../../utils/utils";
+
+const pagingLimit = 2;
 
 function Questions() {
   const { data: sessionData } = useSession();
   const allowedToModify = sessionData?.user.role === "MAINTAINER";
-  const getAllQuery = api.useContext().question.getAll;
 
   const [viewQns, setViewQns] = useState<QuestionMap>(new Map());
   const [changedQns, setChangedQns] = useState(new Set<string>());
   const [deletedQns, setDeletedQns] = useState(new Set<string>());
 
-  const questions = makeMap(
-    api.question.getAll.useQuery(undefined, {
+  const [page, setPage] = useState(0);
+
+  const getAllQuery = api.useContext().question.getAllReducedInfinite;
+
+  const {
+    data,
+    fetchNextPage,
+  } = api.question.getAllReducedInfinite.useInfiniteQuery(
+    {
+      limit: pagingLimit,
+      // titleFilter: sentFilter === "" ? undefined : sentFilter,
+    },
+    {
+      getNextPageParam: (prevPage) => prevPage.nextCursor,
       onSuccess: (data) => {
-        const mappedData = data.map((q) => ({
+        const mappedData = data?.pages[page]?.items.map((q) => ({
           ...(changedQns.has(q.id) ? viewQns.get(q.id) ?? q : q),
         }));
-        setViewQns(makeMap(mappedData, "id"));
+        mappedData && setViewQns(makeMap(mappedData, "id"));
       },
-      onError: (e) => {
+      onError: () => {
         toast.error("Failed to fetch questions");
       },
-    }).data ?? [],
-    "id",
+    },
   );
 
+  const questions = makeMap(data?.pages[page]?.items ?? [], "id");
+
+  useEffect(() => {
```

Comment:
```tsx
  useEffect(() => {
    const mappedData = data?.pages[page]?.items.map((q) => ({
      ...(changedQns.has(q.id) ? viewQns.get(q.id) ?? q : q),
    }));
    mappedData && setViewQns(makeMap(mappedData, "id"));
  }, [page]);
```
I think its better to do this than to invalidate the entire query

Reply:
May have to invalidate on update, delete and etc tho, but this worked for me without doing so

Reply:
May need more testing

Reply:
just realised another dependency for the array should be the data.pages

Reply:
> ```tsx
>   useEffect(() => {
>     const mappedData = data?.pages[page]?.items.map((q) => ({
>       ...(changedQns.has(q.id) ? viewQns.get(q.id) ?? q : q),
>     }));
>     mappedData && setViewQns(makeMap(mappedData, "id"));
>   }, [page]);
> ```
> 
> I think its better to do this than to invalidate the entire query

@Gabau This causes a race condition where the setViewQns is being called before the query is refetched.

https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g39/assets/20167059/c051e748-70e4-4de1-a24d-e8f6461f4744

Reply:
> just realised another dependency for the array should be the data.pages

This dependency addition doesn't help?

Reply:
> > just realised another dependency for the array should be the data.pages
> 
> This dependency addition doesn't help?

If not then ig only way to deal with it is to have a loading state

Reply:
Iirc there is a fetching state we cld use in the infinite query

Reply:
Nvm fixed it, turns out it was just contending with the other handler in `onSuccess` O_O